### ISSUE TITLE
Expansion of a generic type with Promise not working in VsCode (and in PlayGroung) 'type X<T> = () => T | Promise<T>;'

### ISSUE THREAD

**Issue by Luxcium** [microsoft/TypeScript#38760]
**TypeScript Version:**  3.9.2 or 4.0.0-dev.20200523 or  3.3.3333 

**Code**

```ts
export const TypeScript = "v3.9.2";

export class MaybeList<T> {
  public constructor(public value: T) {}
}

// type T_<T> = T
export type T_<T> = T;
// type M_<T> = MaybeList<T>
export type M_<T> = T_<MaybeList<T>>;
// type P_<T> = Promise<T>
export type P_<T> = T_<Promise<T>>;
// type MP_<T> = MaybeList<Promise<T>>
export type MP_<T> = M_<P_<T>>;
// type PM_<T> = Promise<MaybeList<T>>
export type PM_<T> = P_<M_<T>>;
// type PMP_<T> = Promise<MaybeList<Promise<T>>>
export type PMP_<T> = P_<M_<P_<T>>> | P_<MP_<T>> | PM_<P_<T>>;
// type P_T<T> = T | Promise<T>
export type P_T<T> =  T | P_<T>;
// type P_M_<T> = MaybeList<T> | Promise<MaybeList<T>>
export type P_M_<T> = (M_<T> | PM_<T>) | P_T<M_<T>>;
// type P_MP_<T> = MaybeList<Promise<T>> | Promise<MaybeList<Promise<T>>>
export type P_MP_<T> = (MP_<T> | PMP_<T>) | P_T<MP_<T>>;

// type _T_<T> = () => T
export type _T_<T> = () => T_<T_<T>>;
// type _M_<T> = () => MaybeList<T>
export type _M_<T> = () => T_<M_<T>>;
// type _P_<T> = () => Promise<T>
export type _P_<T> = () => T_<P_<T>>;
// type _MP_<T> = () => MaybeList<Promise<T>>
export type _MP_<T> = () => T_<MP_<T>>;
// type _PM_<T> = () => Promise<MaybeList<T>>
export type _PM_<T> = () => T_<PM_<T>>;
// type _PMP_<T> = () => Promise<MaybeList<Promise<T>>>
export type _PMP_<T> = () => T_<PMP_<T>>;
// type _P_T<T> = () => P_T<T>
export type _P_T<T> = () => P_T<T>;
// type _P_M_<T> = () => MaybeList<T> | Promise<MaybeList<T>>
export type _P_M_<T> = () => T_<P_M_<T>>;
// type _P_MP_<T> = () => MaybeList<Promise<T>> | Promise<MaybeList<Promise<T>>>
export type _P_MP_<T> = () => T_<P_MP_<T>>;
```

**Expected behavior:**
I would like to get it expanded like this: `type _P_T<T> = () => T | Promise<T>`

should be similar to: `type _PMP_<T> = () => Promise<MaybeList<Promise<T>>>`
![image](https://user-images.githubusercontent.com/42672814/82762564-b2191200-9dcf-11ea-9e78-405fd2b8c4aa.png)

or should be similar to: `type P_T<T> = T | Promise<T>`
![image](https://user-images.githubusercontent.com/42672814/82762586-daa10c00-9dcf-11ea-97c6-12822f3e2a2b.png)

**Actual behavior:**
Not expanding the `P_T` into `T | Promise<T>`
![image](https://user-images.githubusercontent.com/42672814/82762544-9a418e00-9dcf-11ea-8df8-c1e4dbc9ccad.png)

the `P_` here is for Promise and all other types I have made in the same module are showing `Promise` Keyword expanded except on the line 46~47 (in the playground link below)
`// expansion: type _P_T<T> = () => P_T<T> should  be: type _P_T<T> = () => T | Promise<T>` 
`export type _P_T<T> = () => P_T<T>;`

**Playground Link:** 
[PlayGround Link v3.9.2](https://www.typescriptlang.org/play?ts=3.9.2#code/PQKhCgAIOyFkCSAVSAZBBhAogOQMpZQwiQYD2ADgJ4BOAlgOYAWALpAJWQBMADL5ACEApgDsAVgEMAtnRGQAarIDGotgGkJAZwkUyDADYBXSAApUhgB5K6hqQEoiRYOHBCLummyVkRmtkioKITwlego2AF5IACIANwBmADoATkSuaIBuFzcPL30tTXgJKgAjIVQ6PwAeJAA+SABvKEgKQxL9OiVIb18WGkMlFjIaE1b2zshYiSMhAC5IJDtGgF9wVfBgYEgWQKEFgH0a+qikV3dhth2gg6PIE6zN7d34Q7q7otLyypYjs9yn65wV7HG5wYplCrVOq1B5bK57AAKwPeCJoZBkmiEvxyFwBiORJ0OqPRlSx0NhePgSNuUTBn0hP2JGLJtVqf1x8KpBJeVWp5I2cOeCKBNMgTNJVTpEO+RzZOM8lOF3OpIv5j05wr5IPFmMl4K+1R1LNZ7IVGrgWpRhxFWtZkAAPmLrbaHWKbcCYQLFfskKKUI6jdjzmahT7RQtXVqKRr9qqQVKDT83gG0cy9fSZdDTZdQ3H3iY8wG40sA2G4571bnLbT9QzeamJdDIw3dQm64HoXLgznrkiLdyC5ai1qS07ff3ZVkvZyfQOlhF6qd5T29rPRSZ54vXh7o89Y3O7vU25mu-8Z3mohvD6Cd9O99XTJuxS2WdnKfsH1eFzdbbvrrHPyfY9DRfWU33PQDryQZ1b0rf8lXXJ9A2ApMTWXd8ELeS8n2g3lyz-VdNQPb9kNrGUO1ZU8OXvCcsMfKCiVo1kCMgD8wzor96iRX06nA+92JBTixyOFi2IvejvxQ24UxJVsyKhNDuww-dEIY3kVLVQV4IA4ij3kxlQKbGS0ykijFLPfimPzHCiR0ycXFAYgiHgZA0EwXACEcWByGoehmDYTheH4YRxGkWQFGUVRIA0bRdAMYwzEsaxbAcaBnIRIQaAxTQ6B8SBKkgJhMqEEoqEgBgaAkEQWCEAATAAaSAADMaCEPYyCa7omAkGgGCERqhkgaZ9BaTLNDysgShYCRZFkBghu6Sgyo65yWCYArxqalgAHcer2KraqGzRxusCQasO2qyCUWxVDO3K5Cauh9CEQoTDWvZojwDqdr26I7AGshIFqoRpnykRVqKyAvq23bWsgba6DWshDDYVq-HoQZ7sa5QjFqub4cRphkbYDoZGmlh7u2SG-NYQohmcwxMUa7xqEaqQyDxpqqDZzK+sasYOk0JhGrx9G6BKFH+sgTQ2g6FRfClg7gGGaWhH0EaWboF7nI6qm9mhn7Wsag7tkBoIssR0aaHG3xTfhompD1qHvthvYCqawwaBESoisOwbLuc8bGplkoxCEQY7fe5qyHVsgEZEeaejximfE0WYvJIJBIYkEoyFiPYWdoRhWEgEQyAplQhpEP31sKc3sspsuK72IXhsgMowaUXG6rBob1eczWXsgFWQ78KqKdB3J7sKXWo4N13EgzhYAAksChgB5AAxJAAHUAEEACU14QPAxQP9f5AQAARLAr5iPfT5P6JGp35Bl-XgBVFB94Pg+95wJAABNSAW9IB71QKgaKCAcBX0algAAGgiI+eBT7rwPs5BAFp0C30atAjAqAP5X2gQAcUEF-SAOB14oHQIgJAt8FjrxXmvH+f8AEICwKgze8AsAHwwMvf+SA94CAQOgIB9VnKb2QDgDhp9N5oLAWKQ+SBMAf1QIfMUH8D4InXgQMBMCKHrxwNAzeB8SFYDgLgJAiRIDQIMZALA8hLFQ34RA5ySBV5gK-u-A+qCD6kHXgiQBpjiHLxQO-VAN8fGCDXugIRqA15yL8eAyB+C96YMalfPecA97EI4SA9BsAqGrz8bE4RojAEv1Xu4nhNicB6LARgZRhiQFcIwIYpAf9GmNSQGglA8iik8NfgQcRsBD4nxIZAEx684CNU-n0rh8jbFtJwNIxpCBmmv3cUwje28f5r3kdUyAH9dH9OqQUkgN9wEkMfnUw5eAt670PlgReaVoDOCAA)

**Comment by Luxcium** [comment]
I was working on a project and I realize that some type may have been simplified but the mater of my request is only regarding the `export type _P_T<T> = () => P_T<T>` (which also for consistency should have been P_T_ instead of P_T the leading underscore denotes the use or not of an arrow function as in _P_T<T>)

**Comment by AviVahl** [comment]
TLDR version:

**Code:**

```ts
type MaybePromise<T> = T | Promise<T>
type AsReturnType<T> = () => MaybePromise<T>
```

**Expected:**

Hover over symbol `AsReturnType`, see `() => T | Promise<T>`.

**Actual:**

Hover over symbol `AsReturnType`, see `() => MaybePromise<T>`.

Playground: https://www.typescriptlang.org/play?ssl=2&ssc=16&pln=2&pc=6#code/C4TwDgpgBAsghiARhACgJwPYFsCWBnCAHgBUA+KAXimKgB8p1t8iyAoUSKAQTwGE4ANgMRwAxgGsS5KgAoAlJXLwkqTLgJSgA

Most of the other cases (unions and whatnot) expanded and de-duped types for hover.

**Comment by Luxcium** [comment]
HO my god tanks (I AM SUCH A COMPLICATED PERSON FOR NO GOOD REASON) I will do an effort to never do something like this again (I feel so ridicule tanks @AviVahl )
> TLDR version:
> 
> **Code:**
> 
> ```ts
> type MaybePromise<T> = T | Promise<T>
> type AsReturnType<T> = () => MaybePromise<T>
> ```
>  [...]

**Comment by RyanCavanaugh** [comment]
I don't understand the wall of text and I don't understand the example. Where is `AsCallback` defined? That's not a built-in; it shouldn't appear in any hovers

**Comment by AviVahl** [comment]
@RyanCavanaugh I had typos when renaming stuff, which broke the example. Check out the current expected/actual (my previous comment).

**Comment by Luxcium** [comment]
> I don't understand the wall of text and I don't understand the example. Where is `AsCallback` defined? That's not a built-in; it shouldn't appear in any hovers

I am not as good as you to explain myself I still would like to be able to express my problem.  I want to be able to work with something like `type T_PT_<T> = T | Promise<T>` (type ` T_PT_` means to me 'Tee' or Promise of 'Tee') but somewhere in my  code the "T" is already `T = Promise<ofSomething>`  where `ofSomething` could be anything that `T` would be  (another "T" different because somewhere else in the code but which should be the same instead of the currently defined "T" is a promise of that something)

**Comment by Luxcium** [comment]
I want to use a function  `valueMap_P` and using generic type to extract the "T" as in `Tx = MaybeList<HMSETResult>>` istead of as in `Ty = Promise<MaybeList<HMSETResult>>` I expect to get a "value or promiseOf" because I use a `promisseOf` function

<img width="850" alt="2020-06_20200624_081305_0001" src="https://user-images.githubusercontent.com/42672814/85554397-d9434900-b5f2-11ea-8a1d-929cdbb690a9.png">


```typescript
export async function promisseOf<V>(value: V | Promise<V>): Promise<V> {
  return Promise.resolve(value).then(async x => x);
}
```

<img width="1311" alt="2020-06_20200624_080302_0001" src="https://user-images.githubusercontent.com/42672814/85553527-09d6b300-b5f2-11ea-88bd-3846907ddf3f.png">
```typescript
const valueMap_P: <Promise<MaybeList<HMSETResult>>>(value: T_PT_<Promise<MaybeList<HMSETResult>>>) => ValueMap_P<Promise<MaybeList<HMSETResult>>>
First provide a value<T> (value or promiseOf) and get a ValueThenMap. Then pass a mapFunction will return a Promise<R> (property) valueMap_P
```

**Comment by Luxcium** [comment]
<img width="1144" alt="2020-06_20200624_114538_0001" src="https://user-images.githubusercontent.com/42672814/85589141-52519900-b611-11ea-94ea-cd3544b286f1.png">

Also, I am not sure how I should interpret this problem: `const step7: Promise<Promise<MaybeList<HMSETResult>>>` how can  I remove the `Promise<Promise<any>>`  right now I must do: `step7.then(async x=>x)` 

> the `async` keyword is optional but I use `typescript-eslint/promise-function-async` which requires any function or method that returns a Promise to be marked async...

**Comment by RyanCavanaugh** [comment]
My understanding here has not improved 😕

**Comment by Luxcium** [comment]
I feel like I should be closing this issue until I can explain myself more clearly.

**Comment by Luxcium** [comment]
I must learn how to be more concise it seems. I will use TypeScript the way it is until I can bring my ideas or explain my issues in a more constructive and helpful manner.
